File: typescript/packages/server/src/ssh/session.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { randomBytes } from 'node:crypto'
import { recordStatus } from '@struktoai/mirage-core/workspace/executor/statement'
import type { Workspace } from '@struktoai/mirage-core/workspace/workspace/workspace'
import type { ServerChannel } from 'ssh2'
import type { WorkspaceEntry, WorkspaceRegistry } from '../registry.ts'
import { ChannelInput, ChannelOutput, Mark, channelStdin, deliver } from './stream.ts'

const AGENT_ID = 'ssh'
const INTERRUPTED = 130
const FALLBACK_PROMPT = 'mirage$ '
const LOGIN_HOME = '/'

const encoder = new TextEncoder()
const decoder = new TextDecoder()

function prompt(cwd: string): string {
  return `mirage:${cwd}$ `
}

export function newSessionId(): string {
  return `ssh_${randomBytes(6).toString('hex')}`
}

/**
 * Whether a typed line asks the interactive shell to leave. mirage
 * contains `exit` within the line it runs, so the shell reads the words
 * instead: `exit` or `exit N` leaves, while `exit 1 2` is bash's "too many
 * arguments" and stays.
 */
export function endsShell(line: string): boolean {
  const words = line
    .trim()
    .split(/\s+/)
    .filter((w) => w !== '')
  return words[0] === 'exit' && words.length <= 2
}

export interface Endpoint {
  address: string
  port: number
}

/** What the server knows about one channel's login. */
export interface ChannelRequest {
  username: string
  command: string | null
  term: string | null
  peer: Endpoint | null
  local: Endpoint | null
}

/**
 * The environment sshd hands a login: who, from where, on what. `HOME` is
 * `/`, where every session starts, and the user is the workspace id, since
 * that is the name the client logged in as.
 */
export function loginEnv(request: ChannelRequest): Record<string, string> {
  const env: Record<string, string> = {
    HOME: LOGIN_HOME,
    USER: request.username,
    LOGNAME: request.username,
  }
  const { peer, local } = request
  if (peer !== null && local !== null) {
    env.SSH_CLIENT = `${peer.address} ${String(peer.port)} ${String(local.port)}`
    env.SSH_CONNECTION = `${peer.address} ${String(peer.port)} ${local.address} ${String(local.port)}`
  }
  if (request.term !== null && request.term !== '') env.TERM = request.term
  return env
}

function shellQuote(value: string): string {
  return `'${value.replaceAll("'", `'\\''`)}'`
}

/**
 * Create the session a channel runs as, under the default profile. The
 * login environment is exported by an unrecorded line, so every name
 * clears the session's `pre_session` gate like any `export` would, and a
 * name the profile already set keeps its value.
 */
export async function openSession(
  ws: Workspace,
  sessionId: string,
  env: Record<string, string> = {},
): Promise<void> {
  await ws.ensureSessionsLoaded()
  const session = ws.createSession(sessionId)
  try {
    const missing = Object.entries(env).filter(([name]) => !(name in session.env))
    if (missing.length === 0) return
    const line = `export ${missing.map(([k, v]) => `${k}=${shellQuote(v)}`).join(' ')}`
    await ws.shell(line, { sessionId, record: false })
  } catch (err) {
    await ws.closeSession(sessionId)
    throw err
  }
}

/**
 * Leave `$?` at 130 after a Ctrl-C, as an interactive bash does: an
 * aborted line is the caller's outcome, and an interactive shell is the
 * caller here, whose outcome for an interrupted foreground line is
 * 128 + SIGINT.
 */
function stampInterrupt(ws: Workspace, sessionId: string): void {
  recordStatus(ws.getSession(sessionId), INTERRUPTED)
}

/**
 * One SSH session channel, run as one mirage session of its own.
 *
 * A command channel (`ssh host cmd`) runs its line and ends, and a shell
 * channel reads lines until `exit` or end of input, printing a prompt when
 * the client asked for a terminal. The session is fresh per channel, as
 * each channel is a fresh process under sshd, so a `cd` never leaks from
 * one channel to another and two channels never wait on each other's
 * lines.
 */
export class ShellChannel {
  private readonly tty: boolean
  private readonly input: ChannelInput
  private readonly output: ChannelOutput
  private running: AbortController | null = null
  private lost = false

  constructor(
    private readonly registry: WorkspaceRegistry,
    private readonly entry: WorkspaceEntry,
    private readonly sessionId: string,
    private readonly channel: ServerChannel,
    private readonly request: ChannelRequest,
  ) {
    this.tty = request.term !== null && request.term !== ''
    this.input = new ChannelInput(channel, this.tty)
    this.output = new ChannelOutput(channel, this.tty)
  }

  /** A signal request from the client, delivered like Ctrl-C. */
  signal(): void {
    this.input.interruptReceived()
  }

  private live(): boolean {
    const wid = this.entry.id
    return this.registry.has(wid) && this.registry.get(wid) === this.entry
  }

  private interrupt(): void {
    this.running?.abort()
  }

  private currentPrompt(): string {
    if (!this.live()) return FALLBACK_PROMPT
    return prompt(this.entry.runner.ws.getSession(this.sessionId).cwd)
  }

  /** Run the channel to its end and return the exit status to report. */
  async serve(): Promise<number> {
    this.input.start()
    const onClose = (): void => {
      this.lost = true
      this.interrupt()
    }
    this.channel.once('close', onClose)
    try {
      if (this.request.command !== null) return await this.run(this.request.command)
      return await this.repl()
    } finally {
      this.channel.off('close', onClose)
      this.input.close()
      if (this.live()) await this.entry.runner.ws.closeSession(this.sessionId)
    }
  }

  /**
   * Run one line and stream its output. Ctrl-C and a dropped connection
   * abort the line where it is; the status is then 130.
   */
  async run(line: string): Promise<number> {
    if (!this.live()) {
      await this.output.write(encoder.encode('mirage: the workspace is gone\n'), true)
      return 1
    }
    const ws = this.entry.runner.ws
    const controller = new AbortController()
    this.running = controller
    this.input.onInterrupt(() => {
      controller.abort()
    })
    try {
      const result = await ws.shell(line, {
        sessionId: this.sessionId,
        stdin: channelStdin(this.input),
        agentId: AGENT_ID,
        signal: controller.signal,
      })
      await deliver(result, this.output)
      return result.exitCode
    } catch (err) {
      if (controller.signal.aborted) {
        if (this.lost) return INTERRUPTED
        if (this.live()) stampInterrupt(ws, this.sessionId)
        if (this.tty) await this.output.write(encoder.encode('^C\n'))
        return INTERRUPTED
      }
      if (this.lost) return 1
      const message = err instanceof Error ? err.message : String(err)
      console.warn(`ssh: line failed on ${this.entry.id}: ${message}`)
      await this.output.write(encoder.encode(`mirage: ${message}\n`), true)
      return 1
    } finally {
      this.input.onInterrupt(null)
      this.running = null
    }
  }

  /** Read and run lines until `exit` or end of input. */
  async repl(): Promise<number> {
    let status = 0
    while (!this.lost) {
      if (this.tty) await this.output.write(encoder.encode(this.currentPrompt()))
      const item = await this.input.readline()
      if (item === Mark.LIMIT) {
        await this.output.write(encoder.encode('mirage: shell input line too long\n'), true)
        return 1
      }
      if (item === Mark.EOF) {
        if (this.tty) await this.output.write(encoder.encode('logout\n'))
        return status
      }
      if (item === Mark.INTERRUPT) {
        if (this.tty) await this.output.write(encoder.encode('^C\n'))
        status = INTERRUPTED
        if (this.live()) stampInterrupt(this.entry.runner.ws, this.sessionId)
        continue
      }
      const line = decoder.decode(item).replace(/[\r\n]+$/, '')
      if (line.trim() === '') continue
      status = await this.run(line)
      if (endsShell(line) || !this.live()) return status
    }
    return status
  }
}

function refuse(channel: ServerChannel, message: string): void {
  channel.stderr.write(`mirage: ${message}\n`)
  channel.exit(1)
  channel.end()
}

/**
 * Serve one session channel: a command, or an interactive shell. The SSH
 * username names the workspace. `started` receives the running channel,
 * so the server can hand it the client's signal requests; a refused
 * channel never reaches it.
 */
export async function handleChannel(
  registry: WorkspaceRegistry,
  channel: ServerChannel,
  request: ChannelRequest,
  started: (shell: ShellChannel) => void,
): Promise<void> {
  if (!registry.has(request.username)) {
    refuse(channel, `no such workspace: ${request.username}`)
    return
  }
  const entry = registry.get(request.username)
  const sessionId = newSessionId()
  try {
    await openSession(entry.runner.ws, sessionId, loginEnv(request))
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err)
    console.warn(`ssh: cannot open a session on ${request.username}: ${message}`)
    refuse(channel, `cannot open a session: ${message}`)
    return
  }
  const shell = new ShellChannel(registry, entry, sessionId, channel, request)
  started(shell)
  const status = await shell.serve()
  channel.exit(status)
  channel.end()
}

/** Refuse a subsystem other than SFTP, in the voice of every refusal. */
export function refuseSubsystem(channel: ServerChannel, name: string): void {
  refuse(channel, `unsupported subsystem: ${name}`)
}

File: python/mirage/server/ssh/session.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import concurrent.futures
import logging
import secrets
import shlex
from collections.abc import Mapping

import asyncssh

from mirage import Workspace
from mirage.server.registry import WorkspaceEntry, WorkspaceRegistry
from mirage.server.ssh.stream import (ChannelInput, ChannelOutput, LoopStdin,
                                      Mark, Send, decode, deliver, encode,
                                      loop_sender)
from mirage.workspace.abort import MirageAbortError
from mirage.workspace.executor.statement import record_status

logger = logging.getLogger(__name__)

AGENT_ID = "ssh"
INTERRUPTED = 130
PROMPT = "mirage:{cwd}$ "
FALLBACK_PROMPT = "mirage$ "
LOGIN_HOME = "/"


def new_session_id() -> str:
    return f"ssh_{secrets.token_hex(6)}"


def ends_shell(line: str) -> bool:
    """Whether a typed line asks the interactive shell to leave.

    mirage contains ``exit`` within the line it runs, so the shell reads
    the words instead: ``exit`` or ``exit N`` leaves, while ``exit 1 2``
    is bash's "too many arguments" and stays.

    Args:
        line (str): the line as typed.
    """
    words = line.split()
    return bool(words) and words[0] == "exit" and len(words) <= 2


def login_env(process: asyncssh.SSHServerProcess[str]) -> dict[str, str]:
    """The environment sshd hands a login: who, from where, on what.

    ``HOME`` is ``/``, where every session starts, and the user is the
    workspace id, since that is the name the client logged in as.

    Args:
        process (asyncssh.SSHServerProcess[str]): the channel's process.

    Returns:
        dict[str, str]: variable name to value.
    """
    user = process.get_extra_info("username")
    env = {"HOME": LOGIN_HOME, "USER": user, "LOGNAME": user}
    peer = process.get_extra_info("peername")
    local = process.get_extra_info("sockname")
    if peer and local:
        env["SSH_CLIENT"] = f"{peer[0]} {peer[1]} {local[1]}"
        env["SSH_CONNECTION"] = f"{peer[0]} {peer[1]} {local[0]} {local[1]}"
    if process.term_type:
        env["TERM"] = process.term_type
    return env


async def open_session(ws: Workspace,
                       session_id: str,
                       env: Mapping[str, str] | None = None) -> None:
    """Create the session a channel runs as, under the default profile.

    The login environment is exported by an unrecorded line, so every
    name clears the session's ``pre_session`` gate like any ``export``
    would, and a name the profile already set keeps its value.

    Args:
        ws (Workspace): the workspace, on its own loop.
        session_id (str): the new session's id.
        env (Mapping[str, str] | None): the login environment.
    """
    await ws.ensure_sessions_loaded()
    session = ws.create_session(session_id)
    missing = {k: v for k, v in (env or {}).items() if k not in session.env}
    if not missing:
        return
    line = "export " + " ".join(f"{k}={shlex.quote(v)}"
                                for k, v in missing.items())
    io = await ws.shell(line, session_id=session_id, record=False)
    if io.exit_code != 0:
        logger.debug("ssh: login env refused on %s: %s", session_id, await
                     io.stderr_str())


async def run_line(ws: Workspace, session_id: str, line: str, stdin: LoopStdin,
                   send: Send) -> int:
    """Run one line as the session and stream its output back.

    Runs on the workspace's loop. The status is read after the streams
    drain, because a streaming command settles it only then.

    Args:
        ws (Workspace): the workspace.
        session_id (str): the session the line runs as.
        line (str): the shell line.
        stdin (LoopStdin): the channel's input.
        send (Send): where output goes.

    Returns:
        int: the line's exit status.
    """
    try:
        io = await ws.shell(line,
                            session_id=session_id,
                            stdin=stdin,
                            agent_id=AGENT_ID)
    except MirageAbortError:
        return INTERRUPTED
    await deliver(io.stdout, io.stderr, send)
    return io.exit_code


async def stamp_interrupt(ws: Workspace, session_id: str) -> None:
    """Leave ``$?`` at 130 after a Ctrl-C, as an interactive bash does.

    An aborted line puts ``$?`` back to what it found, because an
    abandoned invocation is the caller's outcome; an interactive shell
    is the caller here, and its outcome for an interrupted foreground
    line is 128 + SIGINT.

    Args:
        ws (Workspace): the workspace, on its own loop.
        session_id (str): the session the line ran as.
    """
    record_status(ws.get_session(session_id), INTERRUPTED)


class ShellChannel:
    """One SSH session channel, run as one mirage session of its own.

    A command channel (``ssh host cmd``) runs its line and ends, and a
    shell channel reads lines until ``exit`` or end of input, printing a
    prompt when the client asked for a terminal. The session is fresh
    per channel, as each channel is a fresh process under sshd, so a
    ``cd`` never leaks from one channel to another and two channels
    never wait on each other's lines.

    Args:
        registry (WorkspaceRegistry): the daemon's workspaces.
        entry (WorkspaceEntry): the workspace this channel serves.
        session_id (str): the session the channel runs as.
        process (asyncssh.SSHServerProcess[str]): the channel's process.
    """

    def __init__(self, registry: WorkspaceRegistry, entry: WorkspaceEntry,
                 session_id: str,
                 process: asyncssh.SSHServerProcess[str]) -> None:
        self._registry = registry
        self._entry = entry
        self._session_id = session_id
        self._process = process
        # asyncssh's line editor, which echoes and turns Ctrl-C into a
        # break, starts only for a pty with a terminal type, so a pty
        # without one is served as the plain stream it effectively is.
        self._tty = bool(process.term_type)
        self._input = ChannelInput(process)
        self._output = ChannelOutput(process, self._tty)
        self._running: concurrent.futures.Future[int] | None = None
        self._lost = False

    def _live(self) -> bool:
        wid = self._entry.id
        return wid in self._registry and self._registry.get(wid) is self._entry

    def _interrupt(self) -> None:
        if self._running is not None:
            self._running.cancel()

    async def _watch_close(self) -> None:
        await self._process.wait_closed()
        self._lost = True
        self._interrupt()

    def _prompt(self) -> str:
        if not self._live():
            return FALLBACK_PROMPT
        cwd = self._entry.runner.ws.get_session(self._session_id).cwd
        return PROMPT.format(cwd=cwd)

    async def serve(self) -> int:
        """Run the channel to its end.

        Returns:
            int: the exit status to report to the client.
        """
        self._input.start()
        watch = asyncio.create_task(self._watch_close())
        try:
            if self._process.command is not None:
                return await self.run(self._process.command)
            return await self.repl()
        finally:
            watch.cancel()
            await asyncio.wait([watch])
            await self._input.close()
            if self._live():
                runner = self._entry.runner
                await runner.call(runner.ws.close_session(self._session_id))

    async def run(self, line: str) -> int:
        """Run one line on the workspace's loop, streaming its output.

        Ctrl-C and a dropped connection cancel the line where it is.

        Args:
            line (str): the shell line.

        Returns:
            int: the line's status; 130 when it was interrupted.
        """
        if not self._live():
            await self._output.write(b"mirage: the workspace is gone\n",
                                     stderr=True)
            return 1
        runner = self._entry.runner
        loop = asyncio.get_running_loop()
        self._running = asyncio.run_coroutine_threadsafe(
            run_line(runner.ws, self._session_id, line,
                     LoopStdin(self._input, loop),
                     loop_sender(self._output, loop)), runner.loop)
        self._input.on_interrupt(self._interrupt)
        try:
            return await asyncio.wrap_future(self._running)
        except asyncio.CancelledError:
            task = asyncio.current_task()
            if task is not None and task.cancelling():
                raise
            if self._lost:
                return INTERRUPTED
            if self._live():
                await runner.call(stamp_interrupt(runner.ws, self._session_id))
            if self._tty:
                await self._output.write(b"^C\n")
            return INTERRUPTED
        except Exception as exc:
            if self._lost:
                logger.debug("ssh: line ended with the connection: %r", exc)
                return 1
            logger.warning("ssh: line failed on %s: %r", self._entry.id, exc)
            await self._output.write(encode(f"mirage: {exc}\n"), stderr=True)
            return 1
        finally:
            self._input.on_interrupt(None)
            self._running = None

    async def repl(self) -> int:
        """Read and run lines until ``exit`` or end of input.

        Returns:
            int: the status of the last line run.
        """
        status = 0
        while not self._lost:
            if self._tty:
                await self._output.write(encode(self._prompt()))
            item = await self._input.readline()
            if item is Mark.EOF:
                if self._tty:
                    await self._output.write(b"logout\n")
                return status
            if item is Mark.INTERRUPT:
                if self._tty:
                    await self._output.write(b"^C\n")
                status = INTERRUPTED
                continue
            line = decode(item).rstrip("\r\n")
            if not line.strip():
                continue
            status = await self.run(line)
            if ends_shell(line) or not self._live():
                return status
        return status


async def handle_process(registry: WorkspaceRegistry,
                         process: asyncssh.SSHServerProcess[str]) -> None:
    """Serve one session channel: a command, or an interactive shell.

    The SSH username names the workspace. SFTP and scp never reach
    here: asyncssh hands them to the SFTP server.

    Args:
        registry (WorkspaceRegistry): the daemon's workspaces.
        process (asyncssh.SSHServerProcess[str]): the channel's process.
    """
    workspace_id = process.get_extra_info("username")
    if process.subsystem is not None:
        process.stderr.write(
            f"mirage: unsupported subsystem: {process.subsystem}\n")
        process.exit(1)
        return
    if workspace_id not in registry:
        process.stderr.write(f"mirage: no such workspace: {workspace_id}\n")
        process.exit(1)
        return
    entry = registry.get(workspace_id)
    session_id = new_session_id()
    runner = entry.runner
    try:
        await runner.call(
            open_session(runner.ws, session_id, login_env(process)))
    except Exception as exc:
        logger.warning("ssh: cannot open a session on %s: %r", workspace_id,
                       exc)
        process.stderr.write(f"mirage: cannot open a session: {exc}\n")
        process.exit(1)
        return
    status = await ShellChannel(registry, entry, session_id, process).serve()
    process.exit(status)
